lib/logdna/client.rb

```diff
@@ -44,95 +63,134 @@ def process_message(msg, opts = {})
       processed_message
     end
 
-    def schedule_flush
-      start_timer = lambda {
-        sleep(@exception_flag ? @retry_timeout : @flush_interval)
-        flush if @flush_scheduled
-      }
-      Thread.new { start_timer.call }
+    def write_to_buffer(msg, opts)
+      Concurrent::Future.execute({ executor: @work_thread_pool }) { write_to_buffer_sync(msg, opts) }
     end
 
-    def write_to_buffer(msg, opts)
-      if @lock.try_lock
-        processed_message = process_message(msg, opts)
-        new_message_size = processed_message.to_s.bytesize
-        @buffer.push(processed_message)
-        @buffer_byte_size += new_message_size
-        @flush_scheduled = true
-        @lock.unlock
-
-        if @flush_limit <= @buffer_byte_size
+    def write_to_buffer_sync(msg, opts)
+      processed_message = process_message(msg, opts)
+      message_size = processed_message.to_s.bytesize
+
+      @lock.synchronize do
+        total_size = message_size
+        if @buffer.any?
+          total_size += @buffer[-1].size
+        end
+        @buffer.push(Message.new(processed_message, total_size))
+
+        if total_size >= @flush_threshold
+          unschedule_flush
           flush
         else
           schedule_flush
         end
-      else
-        @side_message_lock.synchronize do
-          @side_messages.push(process_message(msg, opts))
-        end
       end
     end
 
-    # This method has to be called with @lock
-    def send_request
-      @side_message_lock.synchronize do
-        @buffer.concat(@side_messages)
-        @side_messages.clear
+    ##
+    # Flushes all logs to LogDNA asynchronously
+    def flush(options = {})
+      Concurrent::Future.execute({ executor: @work_thread_pool }) { flush_sync(options) }
+    end
+
+    ##
+    # Flushes all logs to LogDNA synchronously
+    def flush_sync(options = {})
+      slices = @lock.synchronize do
+        # Slice the buffer into chunks that try to be no larger than @flush_limit. Slice points are found with
+        # a binary search thanks to the structure of @buffer. We are working backwards because it's cheaper to 
+        # remove from the tail of an array instead of the head
```

Comment:
`write_to_buffer_sync` schedules a flush when the buffer reaches the `flush_limit` - what scenario does this cover? it might not be worth the 3 iterations to account for lines added to the buffer between the `flush` schedule and when `flush_sync` acquires the lock

Reply:
I rewrote the section in `write_to_buffer_sync` to directly call `flush_sync` to avoid the overhead of scheduling a task on the same thread pool - so the issue you pointed out is even less likely to happen now.

This rebinning isn't strictly needed when `@flush_size` <= `@request_size` which is the default behavior currently. Keeping them separate allows the end user to independently control how much memory they want to use without packing too much data in a single request: `@flush_size` > `@request_size`.

Reply:
I think things would get more interesting when we surface the maximum number of concurrent requests (defaults to 1 for now). I think all 3 of these variables can be used to optimize in high throughput environments